Docs for automatic compaction (#12569)
apache/druid: docs/ingestion/automatic-compaction.md

You can configure automatic compaction dynamically without restarting Druid.
The automatic compaction system uses the following syntax:

```json
{
    "dataSource": <task_datasource>,
    "ioConfig": <IO config>,
    "dimensionsSpec": <custom dimensionsSpec>,
    "transformSpec": <custom transformSpec>,
    "metricsSpec": <custom metricsSpec>,
    "tuningConfig": <parallel indexing task tuningConfig>,
    "granularitySpec": <compaction task granularitySpec>,
    "skipOffsetFromLatest": <time period to avoid compaction>,
    "taskPriority": <compaction task priority>,
    "taskContext": <task context>
}
```

Most fields in the auto-compaction configuration align with a typical [Druid ingestion spec](../ingestion/ingestion-spec.md).
The following properties only apply to auto-compaction:
* `skipOffsetFromLatest`
* `taskPriority`
* `taskContext`

Since the automatic compaction system provides a management layer on top of manual compaction tasks,
the auto-compaction configuration does not include task-specific properties found in a typical Druid ingestion spec.
The following properties are automatically set by the Coordinator:
* `type`: Set to `compact`.

> * `id`: Generated using the task type, datasource name, interval, and timestamp.

Reviewer: It's helpful to know the id starts with `coordinator-issued`. This is how you can tell the difference between compaction tasks that were issued by the auto-compaction system and those issued directly by a user

* `context`: Set according to the user-provided `taskContext`.

For more details on each of the specs in an auto-compaction configuration, see [Automatic compaction dynamic configuration](../configuration/index.md#automatic-compaction-dynamic-configuration).

### Avoid conflicts with ingestion

The Coordinator compacts segments from newest to oldest. In the auto-compaction configuration, you can set a time period, relative to the end time of the most recent segment, for segments that should not be compacted. Assign this value to `skipOffsetFromLatest`. Note that this offset is not relative to the current time but to the latest segment time. For example, if you want to skip over segments from thirty days prior to the end time of the most recent segment, assign `"skipOffsetFromLatest": "P30D"`.

> Compaction tasks that interfere with ingestion tasks will fail. For example, this occurs when an ingestion task needs to write data to a segment for a time interval locked for compaction. To facilitate the continuance of compaction tasks, consider one of the following strategies:

Reviewer: I think it's worth mentioning in this section that compaction tasks failing some times is not an issue. Continuous failures that prevent compaction from making progress is the problem that might need user intervention.

> Compaction tasks that interfere with ingestion tasks will fail. For example, this occurs when an ingestion task needs to write data to a segment for a time interval locked for compaction. To facilitate the continuance of compaction tasks, consider one of the following strategies:

Reviewer 2: Maybe this can rephrased so that pre-empted compaction tasks don't come across as real failures.

```suggestion
Compaction tasks can sometimes get pre-empted if they interfere with ingestion. For example, this occurs when an ingestion task needs to write data to a segment for a time interval locked for compaction. To facilitate the continuance of compaction tasks, consider one of the following strategies:
```

Author: ```suggestion
Compaction tasks may be interrupted when they interfere with ingestion. For example, this occurs when an ingestion task needs to write data to a segment for a time interval locked for compaction. To facilitate the continuance of compaction tasks, consider one of the following strategies:
```

* Set `skipOffsetFromLatest` to reduce the chance of conflicts between ingestion and compaction.

> * Increase the priority value of compaction tasks relative to ingestion tasks. In the auto-compaction configuration, set `taskPriority` to the desired priority value. See [Lock priority](../ingestion/tasks.md#lock-priority) for the priority values of different task types.

Reviewer: We should call out that increasing the priority means that the other ingestion task will fail. If compaction job is competing with streaming ingestion, this means the streaming ingestion task will lag. This is not ideal in most cases and increasing task priority should only be used in very narrow use cases

Reviewer 2: Yeah, I think we shouldn't advise increasing priority at all. Or atleast under a big warn sign saying "for advanced users only" or something.

Author: ```suggestion
* Increase the priority value of compaction tasks relative to ingestion tasks. Only recommended for advanced users. This approach can cause ingestion jobs to fail or lag. To change the priority of compaction tasks, set `taskPriority` to the desired priority value in the auto-compaction configuration. For details on the priority values of different task types, see [Lock priority](../ingestion/tasks.md#lock-priority).
```

```
druid.coordinator.dutyGroups=["compaction"]
druid.coordinator.compaction.duties=["compactSegments"]
druid.coordinator.compaction.period=PT60S
```

## View automatic compaction statistics

After the Coordinator has initiated auto-compaction, you can view compaction statistics for the datasource, including the number of bytes, segments, and intervals already compacted and those awaiting compaction. The Coordinator also reports the total bytes, segments, and intervals not eligible for compaction in accordance with its [segment search policy](../design/coordinator.md#segment-search-policy-in-automatic-compaction).

In the Druid console, the Datasources view displays auto-compaction statistics. The Tasks view shows the task information for compaction tasks that were triggered by the automatic compaction system.

To get statistics by API, send a [`GET` request](../operations/api-reference.md#get-10) to `/druid/coordinator/v1/compaction/status`. To filter the results to a particular datasource, pass the datasource name as a query parameter to the request—for example, `/druid/coordinator/v1/compaction/status?dataSource=wikipedia`.

## Examples

> The following examples demonstrate potential use cases in which auto-compaction may improve your Druid performance. Note that compaction in itself does not modify the underlying data of the segments. However, you can use manual compaction tasks or the automatic compaction system to modify data to improve query performance. See more details in [Compaction strategies](../ingestion/compaction.md#compaction-strategies). The examples in this section do not change the underlying data.

Reviewer: ```suggestion
The following examples demonstrate potential use cases in which auto-compaction may improve your Druid performance. See more details in [Compaction strategies](../ingestion/compaction.md#compaction-strategies). The examples in this section do not change the underlying data.
``` 

I don't think these 2 sentences are needed. It confused me slightly.

### Change segment granularity

> You have a stream set up to ingest data with `HOUR` segment granularity into the `wikistream` datasource. You notice that your Druid segments are smaller than the [recommended segment size](../operations/segment-optimization.md) of 5 million rows per segment. You wish to automatically compact segments to `DAY` granularity while leaving the latest week of data _not_ compacted as your stream may ingest data over that time period.

Reviewer: ```suggestion
You have a stream set up to ingest data with `HOUR` segment granularity into the `wikistream` datasource. You notice that your Druid segments are smaller than the [recommended segment size](../operations/segment-optimization.md) of 5 million rows per segment. You wish to automatically compact segments to `DAY` granularity while leaving the latest week of data _not_ compacted as your stream consistently has data over that time period coming in.
```

nit: I wonder if we want to mention that this is a problem only if the stream has consistently late arriving data for that time period. If there is late arriving data some of the time, that's not too much of a problem. The system is robust enough to handle that.

Author: will add a note below the example

The following auto-compaction configuration compacts existing `HOUR` segments into `DAY` segments while leaving the latest week of data not compacted:

```json
{
  "dataSource": "wikistream",
  "granularitySpec": {
    "segmentGranularity": "DAY"
  },
  "skipOffsetFromLatest": "P1W",
}
```

> ### Update partitioning scheme

Author: ```suggestion
`skipOffsetFromLatest` applies to cases in which your stream consistently ingests late arriving data. If your stream only occasionally ingests late arriving data, the auto-compaction system robustly compacts your data even without `skipOffsetFromLatest`.

### Update partitioning scheme
```

---
id: automatic-compaction
title: "Automatic compaction"
---

<!--
  ~ Licensed to the Apache Software Foundation (ASF) under one
  ~ or more contributor license agreements.  See the NOTICE file
  ~ distributed with this work for additional information
  ~ regarding copyright ownership.  The ASF licenses this file
  ~ to you under the Apache License, Version 2.0 (the
  ~ "License"); you may not use this file except in compliance
  ~ with the License.  You may obtain a copy of the License at
  ~
  ~   http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing,
  ~ software distributed under the License is distributed on an
  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  ~ KIND, either express or implied.  See the License for the
  ~ specific language governing permissions and limitations
  ~ under the License.
  -->

> In Apache Druid, compaction is a special type of ingestion task that reads data from a Druid datasource and writes it back into the same datasource. A common use case for this is to [optimally size segments](../operations/segment-optimization.md) after the data is ingested in Druid to improve query performance. Automatic compaction, or auto-compaction, refers to the system for automatic execution of compaction tasks managed by the [Druid Coordinator](../design/coordinator.md).

Reviewer 3: ```suggestion
In Apache Druid, compaction is a special type of ingestion task that reads data from a Druid datasource and writes it back into the same datasource. A common use case for this is to [optimally size segments](../operations/segment-optimization.md) after ingestion to improve query performance. Automatic compaction, or auto-compaction, refers to the system for automatic execution of compaction tasks managed by the [Druid Coordinator](../design/coordinator.md).
```

> The frequency of compaction tasks relies on the Coordinator [indexing period](../configuration/index.md#coordinator-operation), configured by `druid.coordinator.period.indexingPeriod`.

Reviewer 3: nit: "configured by" reads awkwardly.  Maybe new sentence: You can configure/specify/control the frequency...

The default indexing period is 30 minutes, meaning that the Coordinator first checks for segments to compact at most 30 minutes from when auto-compaction is enabled.

> This time period affects other Coordinator duties including merge and conversion tasks.

Reviewer 2: Nit: Merge tasks are not in use anymore, so we should remove all mentions of merge tasks.

I am not sure what "conversion" tasks are though, I couldn't find any `Task` implementation or even a `CoordinatorDuty` that corresponds to this. It was probably another task type used earlier that has since been removed. @suneet-s , are you aware of these?

We should probably remove all mentions of these from the docs (maybe in a separate PR).

> Note this time period affects other Coordinator duties including merge and conversion tasks.

Reviewer 3: ```suggestion
This time period affects other Coordinator duties including merge and conversion tasks.
```

> To configure the auto-compaction time period, see [Set frequency of compaction runs](#set-frequency-of-compaction-runs).

Reviewer 2: ```suggestion
To configure the auto-compaction time period without affecting the `indexingPeriod`, see [Set frequency of compaction runs](#set-frequency-of-compaction-runs).
```

Author: ```suggestion
To configure the auto-compaction time period without interfering with `indexingPeriod`, see [Set frequency of compaction runs](#set-frequency-of-compaction-runs).
```

> To configure the frequency of compaction tasks, [create a new duty group for the Coordinator](#set-frequency-of-compaction-runs).

Reviewer 3: Not sure we've defined duty group?

Author: I'll edit to leave the intro more general and define it below.
```suggestion
To configure the auto-compaction time period, see [Set frequency of compaction runs](#set-frequency-of-compaction-runs).
```

> At every indexing period, the Coordinator initiates a [segment search](../design/coordinator.md#segment-search-policy-in-automatic-compaction) to determine eligible segments to compact.

Reviewer 2: Since the frequency can be determined either by the `indexingPeriod` or by the period specified in the custom duty, we should probably reword this.

But I am not sure what would be the best way to communicate this. Below is a suggestion, maybe you can phrase it better.

```suggestion
At every invocation of auto-compaction, the Coordinator initiates a [segment search](../design/coordinator.md#segment-search-policy-in-automatic-compaction) to determine eligible segments to compact.
```

Author: Good suggestion!

--data-raw '{
    "dataSource": "wikipedia",
    "granularitySpec": {
        "segmentGranularity": "DAY"
    }
}'
```

To disable auto-compaction for a datasource, send a [`DELETE` request](../operations/api-reference.md#delete-1) to `/druid/coordinator/v1/config/compaction/{dataSource}`. Replace `{dataSource}` with the name of the datasource for which to disable auto-compaction. For example:

```sh
curl --location --request DELETE 'http://localhost:8081/druid/coordinator/v1/config/compaction/wikipedia'
```

## Configure automatic compaction

You can configure automatic compaction dynamically without restarting Druid.
The automatic compaction system uses the following syntax:

```json
{
    "dataSource": <task_datasource>,
    "ioConfig": <IO config>,
    "dimensionsSpec": <custom dimensionsSpec>,
    "transformSpec": <custom transformSpec>,
    "metricsSpec": <custom metricsSpec>,
    "tuningConfig": <parallel indexing task tuningConfig>,
    "granularitySpec": <compaction task granularitySpec>,
    "skipOffsetFromLatest": <time period to avoid compaction>,
    "taskPriority": <compaction task priority>,
    "taskContext": <task context>
}
```

> Most fields in the auto-compaction configuration align with a typical [Druid ingestion spec](../ingestion/ingestion-spec.md).

Reviewer 3: correlate to > align with? not sure

Author: ```suggestion
Most fields in the auto-compaction configuration correlate to a typical [Druid ingestion spec](../ingestion/ingestion-spec.md).
```

Compaction tasks may be interrupted when they interfere with ingestion. For example, this occurs when an ingestion task needs to write data to a segment for a time interval locked for compaction. If there are continuous failures that prevent compaction from making progress, consider one of the following strategies:
* Set `skipOffsetFromLatest` to reduce the chance of conflicts between ingestion and compaction.
* Increase the priority value of compaction tasks relative to ingestion tasks. Only recommended for advanced users. This approach can cause ingestion jobs to fail or lag. To change the priority of compaction tasks, set `taskPriority` to the desired priority value in the auto-compaction configuration. For details on the priority values of different task types, see [Lock priority](../ingestion/tasks.md#lock-priority).

### Set frequency of compaction runs

> If you want the Coordinator to check for compaction more frequently than its indexing period, create a separate duty group and set its time period in the `coordinator/runtime.properties` file. For example, to change the auto-compaction period to 1 minute:

Author: ```suggestion
If you want the Coordinator to check for compaction more frequently than its indexing period, create a separate group to handle compaction duties.
Set the time period of the duty group in the `coordinator/runtime.properties` file. For example, to change the auto-compaction period to 1 minute:
```

In the Druid console, the Datasources view displays auto-compaction statistics. The Tasks view shows the task information for compaction tasks that were triggered by the automatic compaction system.

To get statistics by API, send a [`GET` request](../operations/api-reference.md#get-10) to `/druid/coordinator/v1/compaction/status`. To filter the results to a particular datasource, pass the datasource name as a query parameter to the request—for example, `/druid/coordinator/v1/compaction/status?dataSource=wikipedia`.

## Examples

The following examples demonstrate potential use cases in which auto-compaction may improve your Druid performance. See more details in [Compaction strategies](../ingestion/compaction.md#compaction-strategies). The examples in this section do not change the underlying data.

### Change segment granularity

> You have a stream set up to ingest data with `HOUR` segment granularity into the `wikistream` datasource. You notice that your Druid segments are smaller than the [recommended segment size](../operations/segment-optimization.md) of 5 million rows per segment. You wish to automatically compact segments to `DAY` granularity while leaving the latest week of data _not_ compacted as your stream consistently has data over that time period coming in.

Reviewer 3: ```suggestion
You have a stream set up to ingest data with `HOUR` segment granularity into the `wikistream` datasource. You notice that your Druid segments are smaller than the [recommended segment size](../operations/segment-optimization.md) of 5 million rows per segment. You wish to automatically compact segments to `DAY` granularity while leaving the latest week of data _not_ compacted because your stream consistently receives data within that time period.
```
nit: avoid "as" for "because"

I think this explains the use case for skip offset from latest pretty well. Wonder if there's a way to either refer to this or move it.

Author: updated in later commit

The following auto-compaction configuration compacts existing `HOUR` segments into `DAY` segments while leaving the latest week of data not compacted:

```json
{
  "dataSource": "wikistream",
  "granularitySpec": {
    "segmentGranularity": "DAY"
  },
  "skipOffsetFromLatest": "P1W",
}
```

> `skipOffsetFromLatest` applies to cases in which your stream consistently ingests late arriving data. If your stream only occasionally ingests late arriving data, the auto-compaction system robustly compacts your data even without `skipOffsetFromLatest`.

Reviewer: nit: This makes it sound like the property only applies if the stream has late arriving data. But I think the point you're trying to make. here is that if you have data that arrives late infrequently, you do not need to be as aggressive with this setting.

```suggestion
To choose`skipOffsetFromLatest`, consider how frequently you expect stream consistently ingests late arriving data. If your stream only occasionally ingests events for a time in the past, the auto-compaction system robustly compacts your data even though data is being ingested outside the `skipOffsetFromLatest` window. For most real time use cases, a setting of a few hours or a day would be reasonable.
```

Perhaps a topic for another time. Writing this comment made me realize `late arriving data` might mean different things to different people.  What we mean here is the timestamp field in the event on the stream vs the avg lag it takes to read the data.

> `skipOffsetFromLatest` applies to cases in which your stream consistently ingests late arriving data. If your stream only occasionally ingests late arriving data, the auto-compaction system robustly compacts your data even without `skipOffsetFromLatest`.

Reviewer: nit: This makes it sound like the property only applies if the stream has late arriving data. But I think the point you're trying to make here is that if. you have data that arrives latte infrequently, you do not need to be as aggressive with this setting.

```suggestion
To set choose `skipOffsetFromLatest`, consider how frequently you expect  the stream to be receiving late arriving data. If your stream only occasionally receives late arriving data, the auto-compaction system robustly compacts your data even though data is being ingested outside the `skipOffsetFromLatest` window. For most realtime streaming ingestion use cases, a setting of a few hours or a day would be reasonable.
```
